Core/include/Acts/TrackFitting/KalmanFitter.hpp

```diff
@@ -517,61 +518,21 @@ class KalmanFitter {
-
-        // add a full TrackState entry multi trajectory
-        // (this allocates storage for all components, we will set them later)
-        result.lastTrackIndex = result.fittedStates.addTrackState(
-            TrackStatePropMask::All, result.lastTrackIndex);
-
-        // now get track state proxy back
-        auto trackStateProxy =
-            result.fittedStates.getTrackState(result.lastTrackIndex);
-
-        trackStateProxy.setReferenceSurface(surface->getSharedPtr());
-
-        // assign the source link to the track state
-        trackStateProxy.setUncalibrated(sourcelink_it->second);
+        // do the kalman update
+        auto trackStateProxyRes = detail::kalmanHandleMeasurement(
+            state, stepper, extensions, *surface, sourcelink_it->second,
+            result.fittedStates, result.lastTrackIndex, false);
 
-        // Fill the track state
-        trackStateProxy.predicted() = std::move(boundParams.parameters());
-        if (boundParams.covariance().has_value()) {
-          trackStateProxy.predictedCovariance() =
-              std::move(*boundParams.covariance());
+        if (!trackStateProxyRes.ok()) {
+          return trackStateProxyRes.error();
         }
-        trackStateProxy.jacobian() = std::move(jacobian);
-        trackStateProxy.pathLength() = std::move(pathLength);
 
-        // We have predicted parameters, so calibrate the uncalibrated input
-        // measuerement
-        extensions.calibrator(state.geoContext, trackStateProxy);
+        const auto& trackStateProxy = *trackStateProxyRes;
+        result.lastTrackIndex = trackStateProxy.index();
 
-        // Get and set the type flags
-        auto& typeFlags = trackStateProxy.typeFlags();
-        typeFlags.set(TrackStateFlag::ParameterFlag);
-        if (surface->surfaceMaterial() != nullptr) {
-          typeFlags.set(TrackStateFlag::MaterialFlag);
-        }
-
-        // Check if the state is an outlier.
-        // If not, run Kalman update, tag it as a
-        // measurement and update the stepping state. Otherwise, just tag it as
-        // an outlier
-        if (not extensions.outlierFinder(trackStateProxy)) {
-          // Run Kalman update
-          auto updateRes = extensions.updater(state.geoContext, trackStateProxy,
-                                              state.stepping.navDir, logger);
-          if (!updateRes.ok()) {
-            ACTS_ERROR("Update step failed: " << updateRes.error());
-            return updateRes.error();
-          }
-          // Set the measurement type flag
-          typeFlags.set(TrackStateFlag::MeasurementFlag);
+        // Update the stepper if it is not an outlier
+        if (trackStateProxy.typeFlags().test(
+                Acts::TrackStateFlag::MeasurementFlag)) {
           // Update the stepping state with filtered parameters
           ACTS_VERBOSE("Filtering step successful, updated parameters are : \n"
                        << trackStateProxy.filtered().transpose());
```

Comment:
This log message is then not shared via `kalmanHandleMeasurement`?

Reply:
I thought it might be better leave the decision what is logged to the caller of this, but this obviously could be changed...